src/gtest/test_libzendoo.cpp

```diff
@@ -2491,3 +2501,578 @@ TEST(CctpLibrary, TestCustomFieldsValidationFix)
+
+    // forward transfers: reaching almost the limit
+    for (int i = 0; i < intialFT; i++)  // 4092
+        mtx.vft_ccout.push_back(CTxForwardTransferOut(sidechainId, CAmount(43), uint256S("abba101"), uint160S("abba101")));
+    // Also add a FT for another sidechain
+        mtx.vft_ccout.push_back(CTxForwardTransferOut(sidechainId2, CAmount(43), uint256S("abba101"), uint160S("abba101")));
+
+    CTransaction tx(mtx);
+    ASSERT_TRUE(guardObj.add(tx));
+
+    // Each transaction will go beyond the limit, so we will try to restore the guard state as before the failure
+    for (int i = 4; i < 15; i++) {
+        CMutableTransaction mtx2;
+        mtx2.nVersion = SC_TX_VERSION;
+        mtx2.vsc_ccout.resize(0);
+        mtx2.vft_ccout.resize(0);
+        mtx2.vmbtr_out.resize(0);
+        mtx2.vcsw_ccin.resize(0);
+
+        for (int j = 0; j < i; j++)
+            mtx2.vft_ccout.push_back(CTxForwardTransferOut(sidechainId, CAmount(43), uint256S("abba101"), uint160S("abba101")));
+
+        CTransaction tx2(mtx2);
+
+        // Automatically restore guardObj to a valid state
+        ASSERT_FALSE(guardObj.add(tx2, true));
+
+        auto cbsRef  = guardObj.getCBS();
+        ASSERT_EQ(cbsRef.cbsaMap[sidechainId].ft, intialFT);
+        // And the other sc should have been left intact
+        ASSERT_EQ(cbsRef.cbsaMap[sidechainId2].ft, 1);
+    }
+}
+
+TEST(CctpLibrary, CommitmentBuilder_cleanCBSafterrewind)
+{
+    SidechainTxsCommitmentGuard guardObj;
+
+    uint256 sidechainId = uint256S("abc");
+    uint256 sidechainId2 = uint256S("abcba");
+    CMutableTransaction mtx, mtx2;
+    mtx.nVersion = SC_TX_VERSION;
+    mtx.vsc_ccout.resize(0);
+    mtx.vft_ccout.resize(0);
+    mtx.vmbtr_out.resize(0);
+    mtx.vcsw_ccin.resize(0);
+    mtx2.nVersion = SC_TX_VERSION;
+    mtx2.vsc_ccout.resize(0);
+    mtx2.vft_ccout.resize(0);
+    mtx2.vmbtr_out.resize(0);
+    mtx2.vcsw_ccin.resize(0);
+
+    for (int i = 0; i < 10; i++)
+        mtx.vft_ccout.push_back(CTxForwardTransferOut(sidechainId, CAmount(43), uint256S("abba101"), uint160S("abba101")));
+
+    // forward transfers for different sidechains, sidechainId2 alone should fail
+    mtx2.vft_ccout.push_back(CTxForwardTransferOut(sidechainId, CAmount(43), uint256S("abba101"), uint160S("abba101")));
+    for (int i = 0; i < CCTP_COMMITMENT_BUILDER_FT_LIMIT + 1; i++)  // 4096
+        mtx2.vft_ccout.push_back(CTxForwardTransferOut(sidechainId2, CAmount(43), uint256S("abba101"), uint160S("abba101")));
+
+    CTransaction tx(mtx);
+    CTransaction tx2(mtx2);
+    ASSERT_TRUE(guardObj.add(tx, true));
+    // Transaction must be rejected and cbs state should be restored
+    ASSERT_FALSE(guardObj.add(tx2, true));
+
+    // Guard object must not contain anything about sidechainId2
+    ASSERT_EQ(guardObj.getCBS().cbsaMap.size(), 1);
+    ASSERT_EQ(guardObj.getCBS().cbsaMap.count(sidechainId),  1);
+    ASSERT_EQ(guardObj.getCBS().cbsaMap.count(sidechainId2), 0);
```

Comment:
I'd add just this additional check to ensure internal counters have been rewinded correctly:

```
auto cbsRef  = guardObj.getCBS();
ASSERT_EQ(cbsRef.cbsaMap[sidechainId].ft, 10);
```

Reply:
Check added.